types/react-router/index.d.ts

```diff
@@ -83,9 +83,10 @@ export interface RouteChildrenProps<
 
 export interface RouteProps {
   location?: H.Location;
-  component?: React.ComponentType<RouteComponentProps<any>> | React.ComponentType<any>;
-  render?: ((props: RouteComponentProps<any>) => React.ReactNode);
-  children?: ((props: RouteChildrenProps<any>) => React.ReactNode) | React.ReactNode;
+  // Use React.ComponentType to prevent weak type detection because no properties are supplied
+  component?: React.ComponentType<RouteComponentProps> | React.ComponentType;
+  render?: ((props: RouteComponentProps) => React.ReactNode);
+  children?: ((props: RouteChildrenProps) => React.ReactNode) | React.ReactNode;
```

Comment:
Rather, the `<any>` should be lifted into being a generic argument of the `RouteProps`, and the generic of `Route` fixed 🤔

This is just using `RouteComponentProps` with the default value (which is `{}`), and would reject components that expect more props than what is guaranteed by `{}`.

```jsx
function Component(props: RouteComponentProps<{ foo: string }>) { return null }
<Route path='/:foo' component={Component}/>
```

```
Type '(props: RouteComponentProps<{ foo: string; }, StaticContext, any>) => null' is not assignable to type 'ComponentClass<RouteComponentProps<{}, StaticContext, any>, any> | FunctionComponent<RouteComponentProps<{}, StaticContext, any>> | ComponentClass<{}, any> | FunctionComponent<...> | undefined'.
  Type '(props: RouteComponentProps<{ foo: string; }, StaticContext, any>) => null' is not assignable to type 'FunctionComponent<RouteComponentProps<{}, StaticContext, any>>'.
    Types of parameters 'props' and 'props' are incompatible.
      Type 'RouteComponentProps<{}, StaticContext, any> & { children?: ReactNode; }' is not assignable to type 'RouteComponentProps<{ foo: string; }, StaticContext, any>'.
        Types of property 'match' are incompatible.
          Type 'match<{}>' is not assignable to type 'match<{ foo: string; }>'.
            Property 'foo' is missing in type '{}' but required in type '{ foo: string; }'.
```

Reply:
The route already has a `match` property that contains the items from the path. That being said, there might be a way to set custom `RouteComponentProps` that I don't know of. Otherwise, generics can be removed entirely from its declaration.

Reply:
I'll refine the types more. I don't think it's possible to discover the type of `Params` automatically since that would require parsing of the `path` string.